plugins/image-prioritizer/class-image-prioritizer-background-image-styled-tag-visitor.php

```diff
@@ -65,33 +79,126 @@ public function __invoke( OD_Tag_Visitor_Context $context ): bool {
+	 */
+	private function get_common_lcp_element_external_background_image( OD_URL_Metric_Group $group ): ?array {
+
+		// If the group is not fully populated, we don't have enough URL Metrics to reliably know whether the background image is consistent across page loads.
+		// This is intentionally not using $group->is_complete() because we still will use stale URL Metrics in the calculation.
+		if ( $group->count() !== $group->get_sample_size() ) {
+			return null;
+		}
+
+		$previous_lcp_element_external_background_image = null;
+		foreach ( $group as $url_metric ) {
+			/**
+			 * Stored data.
+			 *
+			 * @var LcpElementExternalBackgroundImage|null $lcp_element_external_background_image
+			 */
+			$lcp_element_external_background_image = $url_metric->get( 'lcpElementExternalBackgroundImage' );
+			if ( ! is_array( $lcp_element_external_background_image ) ) {
+				return null;
+			}
+			if ( null !== $previous_lcp_element_external_background_image && $previous_lcp_element_external_background_image !== $lcp_element_external_background_image ) {
+				return null;
+			}
+			$previous_lcp_element_external_background_image = $lcp_element_external_background_image;
+		}
+
+		return $previous_lcp_element_external_background_image;
+	}
+
+	/**
+	 * Maybe preloads external background image.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param OD_Tag_Visitor_Context $context Context.
+	 */
+	private function maybe_preload_external_lcp_background_image( OD_Tag_Visitor_Context $context ): void {
+		// Gather the tuples of URL Metric group and the common LCP element external background image.
+		if ( ! is_array( $this->group_common_lcp_element_external_background_images ) ) {
+			$this->group_common_lcp_element_external_background_images = array();
+			foreach ( $context->url_metric_group_collection as $group ) {
+				$common = $this->get_common_lcp_element_external_background_image( $group );
+				if ( is_array( $common ) ) {
+					$this->group_common_lcp_element_external_background_images[] = array( $group, $common );
+				}
+			}
+		}
+
+		// There are no common LCP background images, so abort.
+		if ( count( $this->group_common_lcp_element_external_background_images ) === 0 ) {
+			return;
+		}
+
+		$processor = $context->processor;
+		$tag_name  = strtoupper( (string) $processor->get_tag() );
+		foreach ( $this->group_common_lcp_element_external_background_images as $i => list( $group, $common ) ) {
+			if (
+				// Note that the browser may send a lower-case tag name in the case of XHTML or embedded SVG/MathML, but
+				// the HTML Tag Processor is currently normalizing to all upper-case. The HTML Processor on the other
+				// hand may return the expected case.
+				strtoupper( $common['tag'] ) === $tag_name
+				&&
+				$processor->get_attribute( 'id' ) === $common['id'] // May be checking equality with null.
+				&&
+				$processor->get_attribute( 'class' ) === $common['class'] // May be checking equality with null.
```

Comment:
I don't understand why these checks are needed, can you explain please?

Reply:
That makes sense, thanks. That said, I wonder is relying on the web vitals library and its LCP detection the best approach for this functionality? Is there no way to determine the CSS background image for an element by using e.g. `computedStyle` in JS?

Reply:
Hummm. We'd still have to rely on Web Vitals to detect the LCP element regardless. We could use `getComputedStyle` on the LCP element to obtain the CSS background image, but there is a cost to getting the computed style which we can avoid by just using the `url` provided by the `LCPMetric`.

Reply:
That's fair.

It feels a bit fragile to me though because we'd be relying on Web Vitals for something we shouldn't need it for. For the LCP detection, it makes sense to use the library, and I get that it gives us the background image "for free". But what about background images for elements that are not the LCP (for instance if we wanted to lazy-load them)? Thinking this further than only the use-case this PR is addressing, I think it would make sense to solve this in a way that is decoupled from the LCP detection.

Reply:
I'm not sure what's fragile about it? The `LargestContentfulPaint` object has a [`url` property](https://developer.mozilla.org/en-US/docs/Web/API/LargestContentfulPaint/url) which we're reusing.

For background images which are not LCP, lazy-loading them is now being handled as of #1708. However, only background images which are defined inline via `style` attributes are currently lazy-loaded. This is because the tag visitor is able to see that a given element has a background image. For lazy-loading background images which aren't defined with inline `style` attributes, we'd have to compute the style for every single element on the page which would be quite expensive.

Reply:
Ah okay, this would make sense from a performance perspective.

So you're saying we're _not_ going to add lazy-loading for background images that aren't inline? If so, then this makes the current approach more reasonable because all that's in scope is to do it for the LCP element anyway. So then using an approach that works only for the LCP element makes more sense.

Reply:
Correct, I don't envision adding lazy-loading for non-inline background images.